sdks/python/apache_beam/options/pipeline_options.py

```diff
@@ -796,13 +796,13 @@ def _add_argparse_args(cls, parser):
               '"<ENV_VAL>"} }. All fields in the json are optional except '
               'command.'))
     parser.add_argument(
-        '--sdk-worker-parallelism', default=None,
+        '--sdk_worker_parallelism', default=None,
         help=('Sets the number of sdk worker processes that will run on each '
               'worker node. Default is 1. If 0, it will be automatically set '
               'by the runner by looking at different parameters (e.g. number '
               'of CPU cores on the worker machine).'))
     parser.add_argument(
-        '--environment-cache-millis', default=0,
+        '--environment_cache_millis', default=0,
```

Comment:
@goenka, @robertwb, I rootcaused remaining failure on this PR to this typo. 
With this PR, but without fixing this typo, PortableRunner tests fail because the value `environment_cache_millis=10000` set in https://github.com/apache/beam/blob/37556f2cd3ce32e23e16cb653632eafd7169c8e2/sdks/python/build.gradle#L182 is propagated, and then we end up losing correct path to a temporary location stored in the environment while running the test. However I don't fully understand why the test without any changes currently passes. It is possible that another error is lurking somewhere. Do you have any ideas? Also, where is  `environment_cache_millis` actually used? 
cc: @aaltay.

Reply:
This might be an incomplete implementation. I found this still open bug: https://issues.apache.org/jira/browse/BEAM-6853

Reply:
Thanks, left a comment there.

Reply:
Looks like these flags are used in https://github.com/apache/beam/blob/master/runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/ReferenceCountingFlinkExecutableStageContextFactory.java.
We should look how PortableOptions from Python SDK are currently passed to the Flink Java Runner, as something may be off there.

Reply:
@udim pointed me to  the codepath that allows adding options not defined in any of PipelineOptions subclasses: https://github.com/apache/beam/blob/120394be6ca979892bce98bdea6ca01439b22476/sdks/python/apache_beam/options/pipeline_options.py#L227 After looking further, I think I understand what happens:
1) PortableRunner sends a request to get pipeline supported by any given Runner implementing FnAPI: https://github.com/apache/beam/blob/684f8130284a7c7979773300d04e5473ca0ac8f3/sdks/python/apache_beam/runners/portability/portable_runner.py#L251
2) Flink runner replies that `environment_cache_millis` is a supported option.
3) We dynamically try to make `environment_cache_millis` as an option recognizable by arguments the parser in https://github.com/apache/beam/blob/120394be6ca979892bce98bdea6ca01439b22476/sdks/python/apache_beam/runners/portability/portable_runner.py#L281, by specifying add_extra_args_fn.
4)  PortableOptions declares `environment-cache-millis` as an option. Even though the name contains dashes, `argparse` will store the value for this option in a dictionary with a key where dashes are normalized to underscores: https://docs.python.org/dev/library/argparse.html#dest.
5) With this PR, since the value for `environment-cache-millis` is NOT specified on the command line (but a value with underscores is set to 10000), `self._all_options['environment_cache_millis']` is initialized to 0, the current default value. This value takes precedence in `get_all_options()`: https://github.com/apache/beam/blob/120394be6ca979892bce98bdea6ca01439b22476/sdks/python/apache_beam/options/pipeline_options.py#L237, and causes the test to fail.
6) Without this PR, the value for `environment_cache_millis` in `self._all_options` will be initialized to 0 at the first access https://github.com/apache/beam/blob/120394be6ca979892bce98bdea6ca01439b22476/sdks/python/apache_beam/options/pipeline_options.py#L267, but this never happens because `environment_cache_millis` is not used anywhere in Python SDK, so the test passes.

Reply:
Further thoughts on this: 
1. Why is some subset of Flink runner options defined in PortableOptions. It seems that we should remove them, since they are not used by PortableRunner? Any thoughts, @angoenka, @robertwb ?
2. If we need to keep `environment_cache_millis` in `PortableOptions`, is the default 0 reasonable?
3. We should avoid using dashes in pipeline options. Emitting a warning seems to be the safest way, failing or converting them to underscores will be breaking changes.

Reply:
Thanks for the thorough investigation!

Reply:
Some of these could be removed now that there is RunnerOptions. @mxm will have context on this.

Reply:
I think `PortableOptions` should contain options that all Portable Runners are expected to support and `RunnerOptions` may be something that will be obtained from the runner and represent runner-specific options that SDK cannot be aware of. Looks like current `RunnerOptions` class is not used, instead we try to get runner-specific options via JobApi, see https://github.com/apache/beam/blob/e4cf63b9c69cdd8ce37208bb301595f5f2af12c8/sdks/python/apache_beam/runners/portability/portable_runner.py#L281 and above. More changes may be needed to create PipelineOptions subclasses dynamically... Oh well. This is beyond the scope of this PR.

Reply:
@tvalentyn Thanks for taking the time to look at this. I agree that the pipeline options retrieval is not obvious and can cause some confusion. However, before we had to replicate options in all SDKs which proofed to be error-prone.

>  Why is some subset of Flink runner options defined in PortableOptions? It seems that we should remove them, since they are not used by PortableRunner? 

`EnvironmentCacheMillis` is only used by the Flink Runner but it was assumed to be applicable to all portable Runners. I can't speak for Dataflow, but I know that it will be useful for Spark as well. It's used for batch workloads where subsequent operators want to reuse the same environment, e.g. read a local file produced by the preceding operator. I think we could find a better solution for this in the future.

The only other option I see is `sdkWorkerParallelism`. We have experimented with this for the Flink Runner with different work loads and I'm sure other Runners may want to have control over the Harness parallelism.

> If we need to keep environment_cache_millis in PortableOptions, is the default 0 reasonable?

`0` is the default in the Java SDK: https://github.com/apache/beam/commit/63ddad144c1ca7066673b0756c25962f24c899ee#diff-a42abc13cb0431ddc407d59b6023c873R92

>Looks like current RunnerOptions class is not used, instead we try to get runner-specific options via JobApi

`RunnerOptions` is merely a place holder. The `DescribePipelineOptions` request returns all available options from the Java SDK, including their default values.

Reply:
- Thanks, @mxm. I think the meaning of PortableOptions, and the relationship PortableOptions vs RunnerOptions is not obvious and may confuse users. 
- Keeping options common for all/most runners in PortableOptions SGTM. One possibility can be to unify PortableOptions and RunnerOptions in one class (RunnerOptions?) and document that the list of options in this class is not exhaustive. 
- We'd need to add instructions how a user could find which options are supported by a portable runner X.

Reply:
I think we could get rid of RunnerOptions. We created this class just as a dummy when we removed the FlinkOptions from the Python SDK after option fetching had been implemented. 

Not a bad idea to rename rename PortableOptions to RunnerOptions, but we would have to check other SDKs for consistency, e.g. Java has PortablePipelineOptions which are also transferred to the Python SDK.